crypto/keys/src/traits.rs

```diff
@@ -43,6 +44,22 @@ impl From<LengthMismatch32> for KeyError {
     }
 }
 
+/// A trait for an object which is representable as canonical bytes.
+///
+/// This must be general enough to support both X25519Public and RistrettoPublic,
+/// an X25519 object already implements AsRef<[u8]>, but RistrettoPublic cannot,
+/// it must be compressed before you can get a reference to bytes.
+///
+/// This trait attempts to bridge this gap, using a zero-cost abstraction.
+/// An object like X25519 can have GetBytes::Result = &[u8],
+/// and this object already implements AsRef<[u8]>, which can be inlined to a no-op.
+/// An object like Ristretto can have GetBytes::Result = [u8; 32] or similar.
+pub trait GetBytes {
+    type Output: AsRef<[u8]>;
+
+    fn get_bytes(&self) -> Self::Output;
+}
```

Comment:
this is possibly the most controversial part of the whole PR
I introduce this trait because in `trait Kex` we were requiring the `PublicKey` to implement `AsRef<[u8]>` which `RistrettoPublic` simply cannot do -- it must be compressed before it can be viewed as bytes.

This trait is another layer of indirection, so that it can return e.g. `Output = [u8;32]` which itself implements `AsRef<[u8]>`.

It was my hope that `X25519` could keep implementing this with `get_bytes(&self) -> &[u8]` since `&[u8]` implements `AsRef<[u8]>`, but I haven't been able to convince the compiler to build that yet.